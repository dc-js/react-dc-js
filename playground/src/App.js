import React, { useEffect, useState } from 'react'
import crossfilter from 'crossfilter'
import * as d3 from 'd3'
import csv from './ndx.csv'
import { BarChart, PieChart, BubbleChart } from 'react-dc-js'

import 'dc/dist/style/dc.css'
import './App.css'

const numberFormat = d3.format('.2f')

function App() {
  const [cx, setCx] = useState(null)

  useEffect(() => {
    ;(async () => {
      const data = await d3.csv(csv)
      const dateFormatSpecifier = '%m/%d/%Y'
      const dateFormatParser = d3.timeParse(dateFormatSpecifier)
      data.forEach(d => {
        d.dd = dateFormatParser(d.date)
        d.month = d3.timeMonth(d.dd) // pre-calculate month for better performance
        d.close = +d.close // coerce to number
        d.open = +d.open
      })
      const cx = crossfilter(data)
      setCx(cx)
    })()
  }, [])

  if (!cx) {
    return <p>Loading Data...</p>
  }

  const moveMonths = cx.dimension(d => d.month)
  const volumeByMonthGroup = moveMonths
    .group()
    .reduceSum(d => d.volume / 500000)

  const gainOrLoss = cx.dimension(d => (d.open > d.close ? 'Loss' : 'Gain'))
  const gainOrLossGroup = gainOrLoss.group()

  const yearlyDimension = cx.dimension(d => d3.timeYear(d.dd).getFullYear())
  const yearlyPerformanceGroup = yearlyDimension.group().reduce(
    (p, v) => {
      ++p.count
      p.absGain += v.close - v.open
      p.fluctuation += Math.abs(v.close - v.open)
      p.sumIndex += (v.open + v.close) / 2
      p.avgIndex = p.sumIndex / p.count
      p.percentageGain = p.avgIndex ? (p.absGain / p.avgIndex) * 100 : 0
      p.fluctuationPercentage = p.avgIndex
        ? (p.fluctuation / p.avgIndex) * 100
        : 0
      return p
    },
    (p, v) => {
      --p.count
      p.absGain -= v.close - v.open
      p.fluctuation -= Math.abs(v.close - v.open)
      p.sumIndex -= (v.open + v.close) / 2
      p.avgIndex = p.count ? p.sumIndex / p.count : 0
      p.percentageGain = p.avgIndex ? (p.absGain / p.avgIndex) * 100 : 0
      p.fluctuationPercentage = p.avgIndex
        ? (p.fluctuation / p.avgIndex) * 100
        : 0
      return p
    },
    () => ({
      count: 0,
      absGain: 0,
      fluctuation: 0,
      fluctuationPercentage: 0,
      sumIndex: 0,
      avgIndex: 0,
      percentageGain: 0,
    })
  )

  return (
    <div className="App">
      <header className="App-header">
        <BubbleChart
          width={990}
          height={250}
          transitionDuration={1500}
          margins={{ top: 10, right: 50, bottom: 30, left: 40 }}
          dimension={yearlyDimension}
          group={yearlyPerformanceGroup}
          colors={d3.schemeRdYlGn[9]}
          colorDomain={[-500, 500]}
          colorAccessor={d => d.value.absGain}
          keyAccessor={p => p.value.absGain}
          valueAccessor={p => p.value.percentageGain}
          radiusValueAccessor={p => p.value.fluctuationPercentage}
          maxBubbleRelativeSize={0.3}
          x={d3.scaleLinear().domain([-2500, 2500])}
          y={d3.scaleLinear().domain([-100, 100])}
          r={d3.scaleLinear().domain([0, 4000])}
          elasticY={true}
          elasticX={true}
          yAxisPadding={100}
          xAxisPadding={500}
          renderHorizontalGridLines={true}
          renderVerticalGridLines={true}
          xAxisLabel={'Index Gain'}
          yAxisLabel={'Index Gain %'}
          renderLabel={true}
          label={p => p.key}
          renderTitle={true}
          title={p =>
            [
              p.key,
              `Index Gain: ${numberFormat(p.value.absGain)}`,
              `Index Gain in Percentage: ${numberFormat(
                p.value.percentageGain
              )}%`,
              `Fluctuation / Index Ratio: ${numberFormat(
                p.value.fluctuationPercentage
              )}%`,
            ].join('\n')
          }
        />
        <PieChart
          dimension={gainOrLoss}
          group={gainOrLossGroup}
          width={180}
          height={180}
          radius={80}
        />
        <BarChart
          dimension={moveMonths}
          group={volumeByMonthGroup}
          width={990}
          height={40}
          margins={{ top: 0, right: 50, bottom: 20, left: 40 }}
          centerBar={true}
          gap={1}
          x={d3
            .scaleTime()
            .domain([new Date(1985, 0, 1), new Date(2012, 11, 31)])}
          round={d3.timeMonth.round}
          alwaysUseRounding={true}
          xUnits={d3.timeMonths}
        />
      </header>
    </div>
  )
}

export default App
