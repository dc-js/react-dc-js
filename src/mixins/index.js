// @flow

import type { BaseMixinProps } from './base-mixin'
import type { StackMixinProps } from './stack-mixin'

export type MixinProps = BaseMixinProps & StackMixinProps

export { baseMixin } from './base-mixin'
export { stackMixin } from './stack-mixin'
export type { BaseMixinProps, StackMixinProps }
