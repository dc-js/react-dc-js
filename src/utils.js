// @flow

import type { MixinProps } from './mixins'

export function combineMixins(mixins: Array<(any, MixinProps) => any> = []) {
  return function (chart: any, props: MixinProps) {
    // Transforms the received props into corresponding chart methods
    // Ensures that we can map only a subset of props to mixin methods
    const unhandledProps: any = mixins.reduce(
      (value, mixin) => mixin(chart, value),
      props
    )
    // Invoke the remaining generic props to corresponding methods
    Object.keys(unhandledProps).forEach(
      prop => prop in chart && chart[prop](unhandledProps[prop])
    )
  }
}
