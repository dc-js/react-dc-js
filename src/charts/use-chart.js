import { useEffect, useRef } from 'react'
import { InvalidStateException } from 'dc'

import { baseMixin } from '../mixins'
import { combineMixins } from '../utils'

export function useChart(chartFunc, props, mixin = combineMixins([baseMixin])) {
  const chartRef = useRef(null)
  const chart = useRef(null)
  const isRendered = useRef(false)

  useEffect(() => {
    chart.current = chartFunc(chartRef.current)
    isRendered.current = false
  }, [chartFunc])

  useEffect(() => {
    if (!chart.current) return
    mixin && mixin(chart.current, props)
    if (!isRendered.current) {
      try {
        chart.current.render()
        isRendered.current = true
      } catch (e) {
        if (!(e instanceof InvalidStateException)) throw e
      }
    } else {
      chart.current.redraw()
    }
  }, [chartFunc, mixin, props])

  return [chart, chartRef]
}
