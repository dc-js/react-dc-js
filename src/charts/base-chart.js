// @flow

import React from 'react'
import { baseMixin } from '../mixins'
import type { MixinProps } from '../mixins'
import { combineMixins } from '../utils'
import { useChart } from './use-chart'

/**
 * A helper function for generic charts.
 * Allows the hook to remain modular and not return HTML.
 * @param props
 * @param chartFunc
 * @param mixins
 */
export function BaseChart(
  props: $Shape<MixinProps>,
  chartFunc: any => mixed,
  mixins: Array<(any, MixinProps) => mixed> = [baseMixin]
) {
  const [, chartRef] = useChart(chartFunc, props, combineMixins(mixins))

  return <div ref={chartRef} />
}
