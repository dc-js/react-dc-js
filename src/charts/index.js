import {
  barChart,
  boxPlot,
  bubbleChart,
  bubbleOverlay,
  cboxMenu,
  compositeChart,
  dataCount,
  dataTable,
  geoChoroplethChart,
  heatMap,
  htmlLegend,
  legend,
  numberDisplay,
  pieChart,
  rowChart,
  scatterPlot,
  selectMenu,
  seriesChart,
  sunburstChart,
  textFilterWidget,
} from 'dc'
import { baseMixin, stackMixin } from '../mixins'
import { BaseChart } from './base-chart'
import ChartContext from './chart-context'
import { LineChart } from './line-chart'
import { useChart } from './use-chart'

export const BarChart = BaseChart(barChart, [baseMixin, stackMixin])
export const BoxPlot = BaseChart(boxPlot)
export const BubbleChart = BaseChart(bubbleChart)
export const PieChart = BaseChart(pieChart)
export const BubbleOverlay = BaseChart(bubbleOverlay)
export const CboxMenu = BaseChart(cboxMenu)
export const CompositeChart = BaseChart(compositeChart)
export const DataCount = BaseChart(dataCount)
export const DataTable = BaseChart(dataTable)
export const GeoChoroplethChart = BaseChart(geoChoroplethChart)
export const Heatmap = BaseChart(heatMap)
export const HtmlLegend = BaseChart(htmlLegend)
export const Legend = BaseChart(legend)
export const NumberDisplay = BaseChart(numberDisplay)
export const RowChart = BaseChart(rowChart)
export const ScatterPlot = BaseChart(scatterPlot)
export const SelectMenu = BaseChart(selectMenu)
export const SeriesChart = BaseChart(seriesChart)
export const SunburstChart = BaseChart(sunburstChart)
export const TextFilterWidget = BaseChart(textFilterWidget)
export { useChart, LineChart, ChartContext }
