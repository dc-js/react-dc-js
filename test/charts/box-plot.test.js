import React from 'react'
import { render } from '@testing-library/react'
import crossfilter from 'crossfilter'
import * as d3 from 'd3'
import { BoxPlot } from 'src/charts'
import testSet from '../morley.json'

describe('box plot', () => {
  it('should match snapshot', () => {
    const ndx = crossfilter(testSet)
    const dimension = ndx.dimension(d => 'exp-' + d.Expt)
    const group = dimension.group().reduce(
      function (p, v) {
        p.splice(d3.bisectLeft(p, +v.Speed), 0, +v.Speed)
        return p
      },
      function (p, v) {
        p.splice(d3.bisectLeft(p, +v.Speed), 1)
        return p
      },
      function () {
        return []
      }
    )
    const { container } = render(
      <BoxPlot dimension={dimension} group={group} transitionDuration={0} />
    )
    expect(container).toMatchSnapshot()
  })
})
