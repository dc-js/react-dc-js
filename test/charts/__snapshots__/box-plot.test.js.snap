// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`box plot should match snapshot 1`] = `
<div>
  <div
    class="dc-chart"
  >
    <svg
      height="200"
      width="200"
    >
      <g>
        <g
          class="chart-body"
          clip-path="url(http://localhost/#dc-chart1-clip)"
          transform="translate(30, 10)"
        >
          <g
            class="box"
            transform="translate(11.53846153846154, 0)"
          >
            <line
              class="center"
              style="opacity: 1;"
              x1="2.307692307692307"
              x2="2.307692307692307"
              y1="113.93939393939394"
              y2="7.272727272727266"
            />
            <rect
              class="box"
              fill="#1f77b4"
              height="42.02020202020203"
              style="fill-opacity: 1;"
              width="4.615384615384614"
              x="0"
              y="36.36363636363637"
            />
            <line
              class="median"
              x1="0"
              x2="4.615384615384614"
              y1="49.29292929292929"
              y2="49.29292929292929"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="113.93939393939394"
              y2="113.93939393939394"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="7.272727272727266"
              y2="7.272727272727266"
            />
            <circle
              class="outlier"
              cx="2.307692307692307"
              cy="143.03030303030303"
              r="5"
              style="opacity: 0.6;"
            >
              <title>
                650
              </title>
            </circle>
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="78.3838383838384"
            >
              850
            </text>
            <text
              class="box"
              dx="6"
              dy=".3em"
              text-anchor="start"
              x="4.615384615384614"
              y="49.29292929292929"
            >
              940
            </text>
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="36.36363636363637"
            >
              980
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="113.93939393939394"
            >
              740
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="7.272727272727266"
            >
              1,070
            </text>
          </g>
          <g
            class="box"
            transform="translate(34.61538461538461, 0)"
          >
            <line
              class="center"
              style="opacity: 1;"
              x1="2.307692307692307"
              x2="2.307692307692307"
              y1="107.47474747474747"
              y2="42.82828282828282"
            />
            <rect
              class="box"
              fill="#ff7f0e"
              height="27.47474747474746"
              style="fill-opacity: 1;"
              width="4.615384615384614"
              x="0"
              y="67.07070707070707"
            />
            <line
              class="median"
              x1="0"
              x2="4.615384615384614"
              y1="80"
              y2="80"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="107.47474747474747"
              y2="107.47474747474747"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="42.82828282828282"
              y2="42.82828282828282"
            />
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="94.54545454545453"
            >
              800
            </text>
            <text
              class="box"
              dx="6"
              dy=".3em"
              text-anchor="start"
              x="4.615384615384614"
              y="80"
            >
              845
            </text>
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="67.07070707070707"
            >
              885
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="107.47474747474747"
            >
              760
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="42.82828282828282"
            >
              960
            </text>
          </g>
          <g
            class="box"
            transform="translate(57.69230769230769, 0)"
          >
            <line
              class="center"
              style="opacity: 1;"
              x1="2.307692307692307"
              x2="2.307692307692307"
              y1="81.61616161616163"
              y2="58.989898989899"
            />
            <rect
              class="box"
              fill="#2ca02c"
              height="12.929292929292956"
              style="fill-opacity: 1;"
              width="4.615384615384614"
              x="0"
              y="68.68686868686868"
            />
            <line
              class="median"
              x1="0"
              x2="4.615384615384614"
              y1="76.76767676767676"
              y2="76.76767676767676"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="81.61616161616163"
              y2="81.61616161616163"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="58.989898989899"
              y2="58.989898989899"
            />
            <circle
              class="outlier"
              cx="2.307692307692307"
              cy="152.72727272727275"
              r="5"
              style="opacity: 0.6;"
            >
              <title>
                620
              </title>
            </circle>
            <circle
              class="outlier"
              cx="2.307692307692307"
              cy="120.4040404040404"
              r="5"
              style="opacity: 0.6;"
            >
              <title>
                720
              </title>
            </circle>
            <circle
              class="outlier"
              cx="2.307692307692307"
              cy="120.4040404040404"
              r="5"
              style="opacity: 0.6;"
            >
              <title>
                720
              </title>
            </circle>
            <circle
              class="outlier"
              cx="2.307692307692307"
              cy="46.060606060606055"
              r="5"
              style="opacity: 0.6;"
            >
              <title>
                950
              </title>
            </circle>
            <circle
              class="outlier"
              cx="2.307692307692307"
              cy="39.5959595959596"
              r="5"
              style="opacity: 0.6;"
            >
              <title>
                970
              </title>
            </circle>
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="81.61616161616163"
            >
              840
            </text>
            <text
              class="box"
              dx="6"
              dy=".3em"
              text-anchor="start"
              x="4.615384615384614"
              y="76.76767676767676"
            >
              855
            </text>
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="68.68686868686868"
            >
              880
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="81.61616161616163"
            >
              840
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="58.989898989899"
            >
              910
            </text>
          </g>
          <g
            class="box"
            transform="translate(80.76923076923077, 0)"
          >
            <line
              class="center"
              style="opacity: 1;"
              x1="2.307692307692307"
              x2="2.307692307692307"
              y1="120.4040404040404"
              y2="55.757575757575765"
            />
            <rect
              class="box"
              fill="#d62728"
              height="31.51515151515153"
              style="fill-opacity: 1;"
              width="4.615384615384614"
              x="0"
              y="73.53535353535352"
            />
            <line
              class="median"
              x1="0"
              x2="4.615384615384614"
              y1="89.69696969696969"
              y2="89.69696969696969"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="120.4040404040404"
              y2="120.4040404040404"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="55.757575757575765"
              y2="55.757575757575765"
            />
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="105.05050505050505"
            >
              768
            </text>
            <text
              class="box"
              dx="6"
              dy=".3em"
              text-anchor="start"
              x="4.615384615384614"
              y="89.69696969696969"
            >
              815
            </text>
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="73.53535353535352"
            >
              865
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="120.4040404040404"
            >
              720
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="55.757575757575765"
            >
              920
            </text>
          </g>
          <g
            class="box"
            transform="translate(103.84615384615384, 0)"
          >
            <line
              class="center"
              style="opacity: 1;"
              x1="2.307692307692307"
              x2="2.307692307692307"
              y1="113.93939393939394"
              y2="46.060606060606055"
            />
            <rect
              class="box"
              fill="#9467bd"
              height="20.202020202020194"
              style="fill-opacity: 1;"
              width="4.615384615384614"
              x="0"
              y="71.91919191919192"
            />
            <line
              class="median"
              x1="0"
              x2="4.615384615384614"
              y1="91.31313131313132"
              y2="91.31313131313132"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="113.93939393939394"
              y2="113.93939393939394"
            />
            <line
              class="whisker"
              style="opacity: 1;"
              x1="0"
              x2="4.615384615384614"
              y1="46.060606060606055"
              y2="46.060606060606055"
            />
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="92.12121212121211"
            >
              808
            </text>
            <text
              class="box"
              dx="6"
              dy=".3em"
              text-anchor="start"
              x="4.615384615384614"
              y="91.31313131313132"
            >
              810
            </text>
            <text
              class="box"
              dx="-6"
              dy=".3em"
              text-anchor="end"
              x="0"
              y="71.91919191919192"
            >
              870
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="113.93939393939394"
            >
              740
            </text>
            <text
              class="whisker"
              dx="6"
              dy=".3em"
              style="opacity: 1;"
              x="4.615384615384614"
              y="46.060606060606055"
            >
              950
            </text>
          </g>
        </g>
        <g
          class="axis x"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="middle"
          transform="translate(30,170)"
        >
          <path
            class="domain"
            d="M0.5,6V0.5H120.5V6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(13.846153846153847,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              exp-1
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(36.92307692307692,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              exp-2
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(60,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              exp-3
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(83.07692307692308,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              exp-4
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(106.15384615384615,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              exp-5
            </text>
          </g>
        </g>
        <g
          class="axis y"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          text-anchor="end"
          transform="translate(30,10)"
        >
          <path
            class="domain"
            d="M-6,160.5H0.5V0.5H-6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(0,159.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              600
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,143.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              650
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,127.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              700
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,111.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              750
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,95.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              800
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,78.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              850
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,62.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              900
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,46.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              950
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,30.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              1,000
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,14.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              1,050
            </text>
          </g>
        </g>
      </g>
      <defs>
        <clippath
          id="dc-chart1-clip"
        >
          <rect
            height="160"
            transform="translate(-0, -0)"
            width="120"
          />
        </clippath>
      </defs>
    </svg>
  </div>
</div>
`;
